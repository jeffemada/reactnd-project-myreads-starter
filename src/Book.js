import React from 'react';
import PropTypes from 'prop-types';
import ChangeBookshelf from './ChangeBookshelf';
import * as BooksAPI from './BooksAPI';

function Book(props) {
  const { onRefresh, book } = props;

  /**
   * Altera o livro de prateleira.
   * @param {string} shelfId - identificador da pratelira selecionada
   */
  const updateBookshelf = (shelfId) => {
    BooksAPI.update(book, shelfId).then(() => {
      onRefresh();
      // para visualizar imediatamente a nova prateleira na tela de pesquisa
      book.shelf = shelfId;
    });
  };

  return (
    <li key={book.id}>
      <div className="book">
        <div className="book-top">
          <div
            className="book-cover"
            style={
              book.imageLinks && {
                backgroundImage: `url("${book.imageLinks.thumbnail}")`
              }
            }
          />
          <ChangeBookshelf currentShelf={book.shelf ? book.shelf : 'none'} onUpdateBookshelf={updateBookshelf} />
        </div>
        <div className="book-title">{book.title}</div>
        <div className="book-authors">{book.authors ? book.authors.join(', ') : ''}</div>
      </div>
    </li>
  );
}

Book.propTypes = {
  book: PropTypes.object.isRequired,
  onRefresh: PropTypes.func.isRequired
};

export default Book;
