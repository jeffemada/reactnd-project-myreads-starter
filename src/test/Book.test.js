import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import Book from '../Book';

Enzyme.configure({ adapter: new Adapter() });

it('should render a Book', () => {
  const book = {
    title: 'Pro React',
    authors: ['Cassio de Sousa Antonio'],
    imageLinks: {
      thumbnail:
        'http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api'
    },
    id: 'PKpPCwAAQBAJ',
    shelf: 'wantToRead'
  };
  const wrapper = shallow(<Book book={book} onRefresh={jest.fn()} onLoading={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
});

it('should render a Book without authors and imageLinks', () => {
  const book = {
    title: 'Pro React',
    id: 'PKpPCwAAQBAJ'
  };
  const wrapper = shallow(<Book book={book} onRefresh={jest.fn()} onLoading={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
});
