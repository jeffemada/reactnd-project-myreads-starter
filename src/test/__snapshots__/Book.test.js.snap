// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Book 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Book
    book={
      Object {
        "authors": Array [
          "Cassio de Sousa Antonio",
        ],
        "id": "PKpPCwAAQBAJ",
        "imageLinks": Object {
          "thumbnail": "http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
        },
        "shelf": "wantToRead",
        "title": "Pro React",
      }
    }
    onLoading={[MockFunction]}
    onRefresh={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "PKpPCwAAQBAJ",
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="book"
      >
        <div
          className="book-top"
        >
          <div
            className="book-cover"
            style={
              Object {
                "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
              }
            }
          />
          <ChangeBookshelf
            currentShelf="wantToRead"
            onUpdateBookshelf={[Function]}
          />
        </div>
        <div
          className="book-title"
        >
          Pro React
        </div>
        <div
          className="book-authors"
        >
          Cassio de Sousa Antonio
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="book-top"
          >
            <div
              className="book-cover"
              style={
                Object {
                  "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                }
              }
            />
            <ChangeBookshelf
              currentShelf="wantToRead"
              onUpdateBookshelf={[Function]}
            />
          </div>,
          <div
            className="book-title"
          >
            Pro React
          </div>,
          <div
            className="book-authors"
          >
            Cassio de Sousa Antonio
          </div>,
        ],
        "className": "book",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="book-cover"
                style={
                  Object {
                    "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                  }
                }
              />,
              <ChangeBookshelf
                currentShelf="wantToRead"
                onUpdateBookshelf={[Function]}
              />,
            ],
            "className": "book-top",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "book-cover",
                "style": Object {
                  "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentShelf": "wantToRead",
                "onUpdateBookshelf": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Pro React",
            "className": "book-title",
          },
          "ref": null,
          "rendered": "Pro React",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Cassio de Sousa Antonio",
            "className": "book-authors",
          },
          "ref": null,
          "rendered": "Cassio de Sousa Antonio",
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "PKpPCwAAQBAJ",
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="book"
        >
          <div
            className="book-top"
          >
            <div
              className="book-cover"
              style={
                Object {
                  "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                }
              }
            />
            <ChangeBookshelf
              currentShelf="wantToRead"
              onUpdateBookshelf={[Function]}
            />
          </div>
          <div
            className="book-title"
          >
            Pro React
          </div>
          <div
            className="book-authors"
          >
            Cassio de Sousa Antonio
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="book-top"
            >
              <div
                className="book-cover"
                style={
                  Object {
                    "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                  }
                }
              />
              <ChangeBookshelf
                currentShelf="wantToRead"
                onUpdateBookshelf={[Function]}
              />
            </div>,
            <div
              className="book-title"
            >
              Pro React
            </div>,
            <div
              className="book-authors"
            >
              Cassio de Sousa Antonio
            </div>,
          ],
          "className": "book",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="book-cover"
                  style={
                    Object {
                      "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                    }
                  }
                />,
                <ChangeBookshelf
                  currentShelf="wantToRead"
                  onUpdateBookshelf={[Function]}
                />,
              ],
              "className": "book-top",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "book-cover",
                  "style": Object {
                    "backgroundImage": "url(\\"http://books.google.com/books/content?id=PKpPCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api\\")",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "currentShelf": "wantToRead",
                  "onUpdateBookshelf": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Pro React",
              "className": "book-title",
            },
            "ref": null,
            "rendered": "Pro React",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cassio de Sousa Antonio",
              "className": "book-authors",
            },
            "ref": null,
            "rendered": "Cassio de Sousa Antonio",
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`should render a Book without authors and imageLinks 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Book
    book={
      Object {
        "id": "PKpPCwAAQBAJ",
        "title": "Pro React",
      }
    }
    onLoading={[MockFunction]}
    onRefresh={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "PKpPCwAAQBAJ",
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="book"
      >
        <div
          className="book-top"
        >
          <div
            className="book-cover"
          />
          <ChangeBookshelf
            currentShelf="none"
            onUpdateBookshelf={[Function]}
          />
        </div>
        <div
          className="book-title"
        >
          Pro React
        </div>
        <div
          className="book-authors"
        >
          
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="book-top"
          >
            <div
              className="book-cover"
            />
            <ChangeBookshelf
              currentShelf="none"
              onUpdateBookshelf={[Function]}
            />
          </div>,
          <div
            className="book-title"
          >
            Pro React
          </div>,
          <div
            className="book-authors"
          >
            
          </div>,
        ],
        "className": "book",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="book-cover"
              />,
              <ChangeBookshelf
                currentShelf="none"
                onUpdateBookshelf={[Function]}
              />,
            ],
            "className": "book-top",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "book-cover",
                "style": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentShelf": "none",
                "onUpdateBookshelf": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Pro React",
            "className": "book-title",
          },
          "ref": null,
          "rendered": "Pro React",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "book-authors",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "PKpPCwAAQBAJ",
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="book"
        >
          <div
            className="book-top"
          >
            <div
              className="book-cover"
            />
            <ChangeBookshelf
              currentShelf="none"
              onUpdateBookshelf={[Function]}
            />
          </div>
          <div
            className="book-title"
          >
            Pro React
          </div>
          <div
            className="book-authors"
          >
            
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="book-top"
            >
              <div
                className="book-cover"
              />
              <ChangeBookshelf
                currentShelf="none"
                onUpdateBookshelf={[Function]}
              />
            </div>,
            <div
              className="book-title"
            >
              Pro React
            </div>,
            <div
              className="book-authors"
            >
              
            </div>,
          ],
          "className": "book",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="book-cover"
                />,
                <ChangeBookshelf
                  currentShelf="none"
                  onUpdateBookshelf={[Function]}
                />,
              ],
              "className": "book-top",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "book-cover",
                  "style": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "currentShelf": "none",
                  "onUpdateBookshelf": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Pro React",
              "className": "book-title",
            },
            "ref": null,
            "rendered": "Pro React",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "className": "book-authors",
            },
            "ref": null,
            "rendered": "",
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
